Introduce rake and Centroid

gregoryjscott opened the pull request:
This refactors the repository to use `rake` to drive all automation. Centroid and [config.json](https://github.com/ResourceDataInc/please/blob/rake/config.json) keep the Please specific information out of the Rakefile.

This adds necessary files/scripts to make please a NuGet package.

This also removes the NuGet packages (except for NuGet.CommandLine) from the repo. Packages used by the solution will be automatically downloaded by Visual Studio (or even the latest version of Xamarin Studio). The other option is to run `rake install`. `rake install` is required to run other rake tasks.

The rake tasks are shown below.

```
$ rake -T
rake build:debug    # Build for debugging
rake build:release  # Build for release
rake clean:all      # Clean all output
rake clean:release  # Clean release output
rake clean:test     # Clean test output
rake default        # Run tests
rake install        # Install NuGet dependencies
rake release        # Pack and push NuGet package
rake release:pack   # Pack NuGet package
rake release:push   # Push NuGet package
rake test           # Run tests
```

Tasks
- [x] ~~Document details of how everything works in a CONTRIBUTING.md file.~~ This will be handled in a separate PR.

gregoryjscott committed:
Introduce rake and Centroid

gregoryjscott committed:
Delete old scripts and tools

gregoryjscott committed:
Don't ignore gitignores

gregoryjscott committed:
Fix package locations

gregoryjscott committed:
Switch to doing NuGet restores to get packages

gregoryjscott committed:
Ignore packages except for NuGet

gregoryjscott committed:
Remove Bump until it works cross platform

thorncp commented:
The description doesn't say anything about the deleted files. It looks like it's making the switch to pull dependencies from NuGet, instead of packaging them in the repo. Is that the case?

gregoryjscott commented:
> The description doesn't say anything about the deleted files. It looks like it's making the switch to pull dependencies from NuGet, instead of packaging them in the repo. Is that the case?

Yes. I fixed the description.

thorncp commented:
:+1:

thorncp commented on Rakefile:
```diff
@@ -0,0 +1,115 @@
+  desc "Clean release output"
+  task :release do
+    clean(Config.release.output.prep)
+    clean(release_lib);
```
What is this `;` doing here?

gregoryjscott replied:
Stupid pinky finger...

thorncp commented on Rakefile:
```diff
@@ -0,0 +1,115 @@
+  desc "Push NuGet package"
+  task :push do
+    puts "Pushing #{release_nupkg}"
+    # nuget("push #{release_nupkg}")
```
Was this commented out for testing, and then forgotten about?

gregoryjscott committed:
Fix EOFs

gregoryjscott committed:
Remove semicolon

gregoryjscott committed:
Uncomment the release push command

gregoryjscott committed:
Remove optional parenthesis

I asked around and nobody gave me an strong opinion either way with
regards to the convention on parenthesis. I like it with out them so
they're gone now.

gregoryjscott committed:
Support NuGet packages with tools and/or lib files

gregoryjscott committed:
Dont package dependencies
